[#155625342] Create cron process to regularly clean up RDS snapshots

bandesz opened the pull request:
## What

We added a cron job process to the RDS broker to clean up old RDS snapshots. We only delete manual snapshots which have a matching Broker Name AWS tag.

We define a separate job for it so it can run as a separate process.

We migrated from godep to dep as the godep project is not going to be maintained anymore and I also struggled to add a new package as godep complained about a non-existing aws sdk subpackage.

## How to review

Review as part of https://github.com/alphagov/paas-cf/pull/1285

## Who can review

Not me.

Commit:
Implement a cron process for deleting old RDS snapshots

We add a "-cron" flag to the main executable to start a cron process which will regularly delete old RDS snapshots.

We need to delete the old snapshots so we don't reach the AWS limits and also to comply with GDPR requirements.

The following config parameters are needed:
 - cron_schedule: crontab-like expression with seconds precision (e.g. "0 0 * * * *" or "@hourly"), with fields: "second minute hour dom month dow", see [1]
 - keep_snapshots_for_days: only snapshots older than this value will be deleted

We will only delete RDS snapshot which has the "Broker Name" tag matching the broker's name (config: rds_config.broker_name).

[1] https://github.com/robfig/cron/blob/2315d5715e36303a941d907f038da7f7c44c773b/parser.go#L184

Commit:
Migrate from godep to dep

The godep project is not going to be maintained anymore as the project is going
to be archived. [1]

Also I had issues to add a new dependency as godep was complaining about a
non-existing aws sdk subpackage.

We will migrate to dep as it's the most widely supported package manager and we
use it in almost all of our projects.

[1] https://github.com/tools/godep#godep---archived

Commit:
Add github.com/robfig/cron as a dependency

Commit:
Add missing ginkgo bootstrap file to config package

Commit:
Add keep_snapshots_for_days and cron_schedule to the config

alext commented on awsrds/rds_db_instance.go:
```diff
@@ -122,6 +123,42 @@ func (r *RDSDBInstance) DescribeSnapshots(DBInstanceID string) ([]*DBSnapshotDet
+		&rds.DescribeDBSnapshotsInput{},
+		func(page *rds.DescribeDBSnapshotsOutput, lastPage bool) bool {
+			for _, snapshot := range page.DBSnapshots {
+				if snapshot.SnapshotCreateTime.Before(deleteBefore) {
```
If I understand correctly, this will delete all snapshots in the AWS account that are older than N days. I think we should add additional constraints to this so that it only deletes snapshots that are related to instances managed by the broker.

We set a "Broker Name" tag on the instances which looks to get copied to the snapshots. We should filter for snapshots with this tag. (In our deployments, "Broker Name" is set to the `DEPLOY_ENV`)

bandesz replied:
I added filtering for the `Broker Name` tag and we only delete manual snapshots.

Commit:
Document cron job process

Commit:
Move the cron process file to a separate package

Also add a separate configuration file for the cron process so we don't have to
duplicate the huge configuration file in the BOSH release and in the manifest.

Commit:
Only delete manual snapshots

The automated snapshots will be automatically deleted after their retention
period.
